File: frontend/src/services/turnoService.ts
import { get, post, patch, del } from './api';
import type { Turno, CrearTurnoPayload, EstadoTurno } from '../types';

export interface FiltroPeriodo {
  fechaDesde: string; // YYYY-MM-DD
  fechaHasta: string; // YYYY-MM-DD
}

export const turnoService = {
  listar: (estado?: EstadoTurno, periodo?: FiltroPeriodo): Promise<Turno[]> => {
    const params = new URLSearchParams();
    if (estado) params.set('estado', estado);
    if (periodo) {
      params.set('fechaDesde', periodo.fechaDesde);
      params.set('fechaHasta', periodo.fechaHasta);
    }
    const qs = params.toString();
    return get<Turno[]>(`/turnos${qs ? `?${qs}` : ''}`);
  },

  buscarPorId: (id: string): Promise<Turno> =>
    get<Turno>(`/turnos/${id}`),

  listarPorTrabajador: (trabajadorId: string, periodo?: FiltroPeriodo): Promise<Turno[]> => {
    const params = new URLSearchParams();
    if (periodo) {
      params.set('fechaDesde', periodo.fechaDesde);
      params.set('fechaHasta', periodo.fechaHasta);
    }
    const qs = params.toString();
    return get<Turno[]>(`/turnos/trabajador/${trabajadorId}${qs ? `?${qs}` : ''}`);
  },

  crear: (payload: CrearTurnoPayload): Promise<Turno> =>
    post<Turno>('/turnos', payload),

  cambiarEstado: (id: string, estado: EstadoTurno): Promise<Turno> =>
    patch<Turno>(`/turnos/${id}/estado`, { estado }),

  eliminar: (id: string): Promise<void> =>
    del<void>(`/turnos/${id}`),
};

File: applavadero/lib/core/services/turno_service.dart
import '../api/api_client.dart';
import '../api/api_endpoints.dart';
import '../models/turno.dart';

class TurnoService {
  final ApiClient _api;
  TurnoService(this._api);

  Future<List<Turno>> getTurnos(
      {String? fechaDesde, String? fechaHasta, String? estado}) async {
    final params = <String, dynamic>{};
    if (fechaDesde != null) params['fechaDesde'] = fechaDesde;
    if (fechaHasta != null) params['fechaHasta'] = fechaHasta;
    if (estado != null) params['estado'] = estado;

    final res =
        await _api.get(ApiEndpoints.turnos, queryParameters: params);
    return (res as List).map((t) => Turno.fromJson(t)).toList();
  }

  Future<List<Turno>> getTurnosPorTrabajador(
      String trabajadorId, {String? fechaDesde, String? fechaHasta}) async {
    final params = <String, dynamic>{};
    if (fechaDesde != null) params['fechaDesde'] = fechaDesde;
    if (fechaHasta != null) params['fechaHasta'] = fechaHasta;

    final res = await _api.get(
        '/api/turnos/trabajador/$trabajadorId',
        queryParameters: params);
    return (res as List).map((t) => Turno.fromJson(t)).toList();
  }

  Future<Turno> getTurno(String id) async {
    final res = await _api.get(ApiEndpoints.turnoDetalle(id));
    return Turno.fromJson(res as Map<String, dynamic>);
  }

  Future<Turno> crearTurno(Map<String, dynamic> data) async {
    final res = await _api.post(ApiEndpoints.turnos, data: data);
    return Turno.fromJson(res as Map<String, dynamic>);
  }

  Future<Turno> cambiarEstado(String id, String estado) async {
    final res = await _api
        .patch(ApiEndpoints.turnoEstado(id), data: {'estado': estado});
    return Turno.fromJson(res as Map<String, dynamic>);
  }
}
